pkg/controller/render/machine_config_to_ignition_test.go

```diff
@@ -0,0 +1,180 @@
+	require.Nil(t, err)
+	return ign3types.File{
+		Node: ign3types.Node{
+			Path: path,
+		},
+		FileEmbedded1: ign3types.FileEmbedded1{
+			Contents: ign3types.Resource{
+				Source: util.StrToPtr(src),
+			},
+			Mode: util.IntToPtr(0644),
+		},
+	}
+}
+
+func TestMachineConfigToIgnition(t *testing.T) {
+	// common inputs
+	defaultInputIgnition := ign3types.Config{
+		Ignition: ign3types.Ignition{
+			Version: "3.2.0",
+		},
+	}
+	rawDefaultInputIgnition, err := json.Marshal(defaultInputIgnition)
+	require.Nil(t, err)
+	defaultInputConfig := runtime.RawExtension{
+		Raw: rawDefaultInputIgnition,
+	}
+
+	// expected content
+	defaultMCDContent, err := json.Marshal(MCDContent{})
+	require.Nil(t, err)
+	defaultMCDContentFile := helpers.NewIgnFile("/etc/machine-config-daemon/mcd_content.json", string(defaultMCDContent))
+	defaultExpectedIgnition := defaultInputIgnition
+	defaultExpectedIgnition.Storage.Files = append(defaultExpectedIgnition.Storage.Files, defaultMCDContentFile)
+
+	testCases := []machineConfigToIgnitionTestCase{
+		// test OSImageURL is simple enough to be added to testCases directly
+		{
+			name: "test OSImageURL",
+			spec: mcfgv1.MachineConfigSpec{
+				Config:     defaultInputConfig,
+				OSImageURL: "URL that should be ignored",
+			},
+			expectedIgnition: defaultExpectedIgnition,
+		},
+	}
+
+	// test Config
+	inputIgnition := defaultInputIgnition
+	tempUser := ign3types.PasswdUser{Name: "core", SSHAuthorizedKeys: []ign3types.SSHAuthorizedKey{"5678", "abc"}}
+	inputIgnition.Passwd.Users = []ign3types.PasswdUser{tempUser}
+	rawInputIgnition, err := json.Marshal(inputIgnition)
+	require.Nil(t, err)
+
+	expectedIgnition, err := ctrlcommon.ParseAndConvertConfig(rawInputIgnition)
+	expectedIgnition.Storage.Files = append(expectedIgnition.Storage.Files, defaultMCDContentFile)
+	require.Nil(t, err)
+
+	testCases = append(testCases, machineConfigToIgnitionTestCase{
+		name: "test Config",
+		spec: mcfgv1.MachineConfigSpec{
+			Config: runtime.RawExtension{
+				Raw: rawInputIgnition,
+			},
+		},
+		expectedIgnition: expectedIgnition,
+	})
+
+	// test Extensions
+	treeFile := createTreeFile(t, "/etc/rpm-ostree/origin.d/extensions-7103bfc.yaml", []byte(`# Generated by the MCO
+
+packages:
+    - kernel-devel
+    - kernel-headers
+override-remove: []
+`))
+	expectedIgnition = defaultExpectedIgnition
+	expectedIgnition.Storage.Files = append([]ign3types.File{treeFile}, expectedIgnition.Storage.Files...)
+	testCases = append(testCases, machineConfigToIgnitionTestCase{
+		name: "test Extensions",
+		spec: mcfgv1.MachineConfigSpec{
+			Config:     defaultInputConfig,
+			Extensions: []string{"kernel-devel"},
+		},
+		expectedIgnition: expectedIgnition,
+	})
+
+	// test KernelType
+	treeFile = createTreeFile(t, "/etc/rpm-ostree/origin.d/extensions-cd51cd4.yaml", []byte(`# Generated by the MCO
+
+packages:
+    - kernel-rt
+    - kernel-rt-modules
+    - kernel-rt-modules-extra
+override-remove:
+    - kernel
```

Comment:
Side note, this currently requires https://github.com/coreos/rpm-ostree/pull/3501 to start and there will be more work specifically to handle kernel.